javaagent-tooling/src/main/java/io/opentelemetry/javaagent/tooling/ExtensionClassLoader.java

```diff
@@ -0,0 +1,172 @@
+/*
+ * Copyright The OpenTelemetry Authors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package io.opentelemetry.javaagent.tooling;
+
+import static io.opentelemetry.javaagent.tooling.ShadingRemapper.rule;
+
+import java.io.ByteArrayInputStream;
+import java.io.File;
+import java.io.IOException;
+import java.io.InputStream;
+import java.net.MalformedURLException;
+import java.net.URISyntaxException;
+import java.net.URL;
+import java.net.URLClassLoader;
+import java.util.jar.JarFile;
+import java.util.jar.Manifest;
+import net.bytebuddy.jar.asm.ClassReader;
+import net.bytebuddy.jar.asm.ClassWriter;
+import net.bytebuddy.jar.asm.commons.ClassRemapper;
+
+/**
+ * This classloader is used to load arbitrary extensions for Otel Java instrumentation agent. Such
+ * extensions may include SDK components (exporters or propagators) and additional instrumentations.
+ * They have to be isolated and shaded to reduce interference with the user application and to make
+ * it compatible with shaded SDK used by the agent.
+ */
+// TODO find a way to initialize logging before using this class
+public class ExtensionClassLoader extends URLClassLoader {
+
+  // We need to prefix the names to prevent the gradle shadowJar relocation rules from touching
+  // them. It's possible to do this by excluding this class from shading, but it may cause issue
+  // with transitive dependencies down the line.
+  private static final ShadingRemapper remapper =
+      new ShadingRemapper(
+          rule("#io.opentelemetry.api", "#io.opentelemetry.javaagent.shaded.io.opentelemetry.api"),
+          rule(
+              "#io.opentelemetry.context",
+              "#io.opentelemetry.javaagent.shaded.io.opentelemetry.context"),
+          rule(
+              "#io.opentelemetry.extension.aws",
+              "#io.opentelemetry.javaagent.shaded.io.opentelemetry.extension.aws"),
```

Comment:
This is the x-ray propagator - I think it's to be able to inject into user classloader for use from instrumentation, even if they have their own (potentially incompatible) one. Since extension instrumentation may use the propagator it seems appropriate to shade here too.